src/entities/road.lua

```diff
@@ -121,36 +150,37 @@ function road:update(angle, roadShift)
     end
   end
 
-  -- Set angles of all frontier objects to the new angle
-  self.frontier.main.body:setAngle(angle)
-  self.frontier.left.body:setAngle(angle)
-  self.frontier.right.body:setAngle(angle)
-
   -- Get the old frontier position, and calculate the new position based on the
   -- new angle.
-  mainX = self.frontier.main.body:getX()
-  mainY = self.frontier.main.body:getY()
-  deltaMainX = (self.roadPush) * math.cos(angle)
-  deltaMainY = (self.roadPush) * math.sin(angle)
+  local x = self.frontier.main.body:getX() + (self.roadPush) * math.cos(angle)
+  local y = self.frontier.main.body:getY() + (self.roadPush) * math.sin(angle)
 
+  self:setFrontier(x, y, angle, roadShift)
+  return self.frontier.main.body:getX(), self.frontier.main.body:getY(), self.frontier.main.body:getAngle()
+end
+
+function road:setFrontier(x, y, angle, roadShift)
   -- calculate positions of left and right frontiers relative to the new main
   -- frontier.
-  leftAngle = angle - (math.pi/2)
+  local leftAngle = angle - (math.pi/2)
   if leftAngle < -math.pi then
     leftAngle = leftAngle + 2 * math.pi
   end
-  deltaLeftX = self.frontier.main.length/3 * math.cos(leftAngle)
-  deltaLeftY = self.frontier.main.length/3 * math.sin(leftAngle)
-  rightAngle = angle + (math.pi/2)
+  local deltaLeftX = self.frontier.main.length/3 * math.cos(leftAngle)
+  local deltaLeftY = self.frontier.main.length/3 * math.sin(leftAngle)
+  local rightAngle = angle + (math.pi/2)
   if rightAngle > math.pi then
     rightAngle = rightAngle - 2 * math.pi
   end
-  deltaRightX = self.frontier.main.length/3 * math.cos(rightAngle)
-  deltaRightY = self.frontier.main.length/3 * math.sin(rightAngle)
-
+  local deltaRightX = self.frontier.main.length/3 * math.cos(rightAngle)
+  local deltaRightY = self.frontier.main.length/3 * math.sin(rightAngle)
+  
   -- if the player is near the edge of the frontier, calculate deltaSlides,
   -- which shift the frontier perpendicular to the trajectory of the player to
   -- center the frontier infront of the player.
+  local deltaSlideX = nil
+  local deltaSlideY = nil
+  local slideAngle = nil
```

Comment:
I'm having a scoping issue here I can't figure out. When this code is run I get an error on line 193. `math.cos()` is getting a nil value, which makes me think that `slideAngle` at the scope of the call on 193 isn't getting the assignment on lines 189/191. I assumed that since `slideAngle` is being defined at a higher scope the inheritance would work. Any suggestions?

Reply:
`slideAngle` is `nil` there because it is conditionally assigned based on the value of `roadShift`, which is itself `nil`.  when main calls `setFrontier()`, it passes in `nil` for `roadShift`, and so `slideAngle` is never set.  here is the problem in main:

```lua
local roadShift
if leftDistance < paveThreshold then
  local roadShift = "left"
elseif rightDistance < paveThreshold then
  local roadShift = "right"
else
  local roadShift = "center"
end
local x, y, angle = road:pushFrontier(p:getTrajectory(), roadShift)
```

the first `local roadShift` above the `if` declares the variable `roadShift` as local to that current scope, and that's good.  at that scope, the variable can be passed into `pushFrontier()` as you do here.  nevertheless, the assignments in the `if` statements never make it to this declaration.  they are instead going to freshly declared `local RoadShift` variables in the bodies of the `if` statements, which are then immediately thrown away.  if you remove the `local` modifier from the assignments, these won't be declarations for new variables anymore, and the interpreter will instead assign to the `local roadShift` that you declared above the `if`.

more about lua scope here: http://lua-users.org/wiki/ScopeTutorial

Reply:
heh, I caught that issue when I parsed 'roadShift` inside `pushFrontier()` but not when I was assigning `roadShift` in main.

Thanks for the explanation.